src/main/java/com/ibm/etcd/client/config/EtcdClusterConfig.java

```diff
@@ -225,12 +276,26 @@ public static EtcdClusterConfig fromJsonFileOrSimple(String fileOrSimpleString)
         throw new FileNotFoundException("etcd config json file not found: " + f);
     }
 
-    private static final Cache<CacheKey,EtcdClient> clientCache = CacheBuilder.newBuilder().weakValues()
-            .<CacheKey,EtcdClient>removalListener(rn -> rn.getValue().close()).build();
+    private static final Cache<CacheKey, EtcdClient> clientCache = CacheBuilder.newBuilder().weakValues()
+            .<CacheKey, EtcdClient>removalListener(rn -> EtcdClient.Internal.cleanup(rn.getValue())).build();
 
     public static EtcdClient getClient(EtcdClusterConfig config) throws IOException, CertificateException {
         try {
-            return clientCache.get(new CacheKey(config), config::newClient);
+            // Share equivalent ref-counted client from cache if possible
+            final CacheKey key = new CacheKey(config);
+            while (true) {
```

Comment:
Why do we need to loop here?

Reply:
This is in case a matching client found in the cache was already closed... we need to flush it and try to get/set it again.